hare/hare.go

```diff
@@ -0,0 +1,239 @@
+
+	h.lastLayer = 0
+
+	h.outputChan = make(chan TerminationOutput, h.bufferSize)
+	h.outputs = make(map[mesh.LayerID][]mesh.BlockID, h.bufferSize) //  we keep results about LayerBuffer past layers
+
+	h.factory = func(conf config.Config, key crypto.PublicKey, instanceId InstanceId, s *Set, oracle Rolacle, signing Signing, p2p NetworkService, terminationReport chan TerminationOutput) Consensus {
+		return NewConsensusProcess(conf, key, instanceId, s, oracle, signing, p2p, terminationReport)
+	}
+
+	return h
+}
+
+func (h *Hare) isTooLate(id mesh.LayerID) bool {
+	h.layerLock.RLock()
+	if int(id) < int(h.lastLayer)-h.bufferSize {
+		h.layerLock.RUnlock()
+		return true
+	}
+	h.layerLock.RUnlock()
+	return false
+}
+
+// ErrTooLate means that the consensus was terminated too late
+var ErrTooLate = errors.New("consensus process %v finished too late")
+
+func (h *Hare) collectOutput(output TerminationOutput) error {
+	id := common.BytesToUint32(output.Id())
+
+	if h.isTooLate(mesh.LayerID(id)) {
+		return ErrTooLate
+	}
+
+	set := output.Set()
+	blocks := make([]mesh.BlockID, len(set.values))
+	i := 0
+	for _, v := range set.values {
+		blocks[i] = mesh.BlockID(common.BytesToUint32(v.Bytes()))
+		i++
+	}
+	h.mu.Lock()
+	if len(h.outputs) == h.bufferSize {
+		for k := range h.outputs {
+			if int(k) < int(h.lastLayer)-h.bufferSize {
+				delete(h.outputs, k)
+			}
+		}
+	}
+	h.outputs[mesh.LayerID(id)] = blocks
+	h.mu.Unlock()
+
+	return nil
+}
+
+func (h *Hare) onTick(id mesh.LayerID) {
+	h.layerLock.Lock()
+	if id > h.lastLayer {
+		h.lastLayer = id
+	}
+	h.layerLock.Unlock()
+
+	ti := time.NewTimer(h.networkDelta)
+	select {
+	case <-ti.C:
+		break // keep going
+	case <-h.CloseChannel():
+		// closed while waiting the delta
+		return
+	}
+
+	// retrieve set form orphan blocks
+	blocks := h.obp.GetOrphanBlocks()
+
+	set := NewEmptySet(len(blocks))
+
+	for _, b := range blocks {
+		// todo: figure out real type of blockid
+		set.Add(Value{NewBytes32(b.ToBytes())})
+	}
+
+	instid := InstanceId{NewBytes32(id.ToBytes())}
+
+	cp := h.factory(h.config, h.pubKey, instid, set, h.rolacle, h.sign, h.network, h.outputChan)
+	cp.Start()
+	h.broker.Register(cp)
+}
+
+var (
+	// ErrTooOld is an error we return when we've been requested output about old consensus procs
+	ErrTooOld = errors.New("results for that layer already deleted")
+	// ErrTooEarly is what we return when the requested layer consensus is still in process
+	ErrTooEarly = errors.New("results for that layer haven't arrived yet")
+)
+
+// GetResults returns the hare output for a given LayerID. returns error if we don't have results yet.
+func (h *Hare) GetResult(id mesh.LayerID) ([]mesh.BlockID, error) {
+	if h.isTooLate(id) {
+		return nil, ErrTooOld
+	}
+
+	h.mu.RLock()
+	blks, ok := h.outputs[id]
+	if !ok {
+		h.mu.RUnlock()
+		return nil, ErrTooEarly
+	}
+	h.mu.RUnlock()
+	return blks, nil
+}
+
+func (h *Hare) outputCollectionLoop() {
+	// todo: think about adding some concurrency here.
```

Comment:
Don't add and remove the comment. We don't expect many outputs at the same time and even if we did, collectOutput is very simple and quick.